### PR TITLE
(Archiving) Add retrieval functionality

### PR DESCRIPTION
**Pull request by islean**
## Description

This PR adds CLI functionality for retrieving samples, cases and orders from DDN.

### Added

- CLI command for retrieving a sample
- CLI command for retrieving a case
- CLI command for retrieving an order

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b 3182-add-retrieve-commands -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by islean** [comment]
Adding some tests

File: cg/cli/archive.py [hunk 1]
```diff
@@ -82,3 +82,48 @@ def delete_file(context: CGConfig, dry_run: bool, file_path: str):
         data_flow_config=context.data_flow,
     )
     spring_archive_api.delete_file(file_path=file_path, dry_run=dry_run)
+
+
+@archive.group("retrieve-spring")
```
Comment (Vince-janv) [inline cg/cli/archive.py:87]: This was just my working title. The whole command will be quite long:
`cg archive retrieve-spring case justhusky`
Do you have any suggestions?

File: cg/meta/archive/archive.py [hunk 2]
```diff
@@ -328,3 +329,16 @@ def delete_file(self, file_path: str, dry_run: bool = False) -> None:
             file_and_sample=file_and_sample, archive_location=archive_location
         )
         self.housekeeper_api.delete_file(file.id)
+
+    def retrieve_spring_files_for_order(self, ticket_id) -> None:
```
Comment (Vince-janv) [inline cg/meta/archive/archive.py:333]: int or str?
```suggestion
    def retrieve_spring_files_for_order(self, ticket_id: int) -> None:
```

File: cg/meta/archive/archive.py [hunk 3]
```diff
@@ -328,3 +329,16 @@ def delete_file(self, file_path: str, dry_run: bool = False) -> None:
...
+        order: Order = self.status_db.get_order_by_ticket_id(ticket_id)
+        for case in order.cases:
+            self.retrieve_spring_files_for_case(case.internal_id)
+
+    def retrieve_spring_files_for_sample(self, sample_id) -> None:
```
Comment (Vince-janv) [inline cg/meta/archive/archive.py:338]:
```suggestion
    def retrieve_spring_files_for_sample(self, sample_id: str) -> None:
```

File: cg/cli/archive.py [hunk 4]
```diff
@@ -82,3 +82,48 @@ def delete_file(context: CGConfig, dry_run: bool, file_path: str):
...
+def retrieve_spring():
+    """Retrieve spring files."""
+    pass
+
+
+@retrieve_spring.command("sample")
+@click.pass_obj
+@click.argument("sample_id", required=True)
+def retrieve_spring_sample(context: CGConfig, sample_id: str):
```
Comment (Vince-janv) [inline cg/cli/archive.py:96]: Maybe all of these should have `for` in them. What do you think?
```suggestion
def retrieve_spring_for_sample(context: CGConfig, sample_id: str):
```

**Review by Vince-janv — APPROVED** [review]
Lightning fast! ⚡

File: cg/cli/archive.py (hunk 1 above)
Comment (islean) [reply cg/cli/archive.py:87]: Maybe just `retrieve` then?

File: cg/cli/archive.py (hunk 4 above)
Comment (islean) [reply cg/cli/archive.py:96]: Makes sense given the names in the SpringArchiveAPI as well 💪

File: cg/cli/archive.py [hunk 5]
```diff
@@ -82,3 +82,48 @@ def delete_file(context: CGConfig, dry_run: bool, file_path: str):
...
+        housekeeper_api=context.housekeeper_api,
+        data_flow_config=context.data_flow,
+    )
+    spring_archive_api.retrieve_spring_files_for_case(case_id=case_id)
+
+
+@retrieve_spring.command("ticket")
+@click.pass_obj
+@click.argument("ticket_id", required=True)
+def retrieve_spring_files_for_ticket(context: CGConfig, ticket_id: int):
```
[the remark this answers is not in the document]
Comment (Vince-janv) [reply cg/cli/archive.py:122]: Production will definitely want `-t <ticket_id>` which I think makes sense for how cg looks right now. But retrieving a ticket feels wrong. To me `cg retrieve order <order_id>` strikes a good balance

[the remark this answers is not in the document]
Comment (seallard) [reply cg/cli/archive.py:122]: Are these CLI commands only for manual use? If so, maybe we can start relying on orders instead.

[the remark this answers is not in the document]
Comment (seallard) [reply cg/cli/archive.py:122]: I don't see any problem with using order ids here 👍

[the remark this answers is not in the document]
Comment (Vince-janv) [reply cg/cli/archive.py:122]: Should we generate an internal_id for orders like we do with cases? Might be good if we want them memorised and read by humans

[the remark this answers is not in the document]
Comment (islean) [reply cg/cli/archive.py]: Something like the applied modification then? Not sure how to write it cleanly though..

[the remark this answers is not in the document]
Comment (islean) [reply cg/cli/archive.py]: I think I am slightly in favour of some internal_id but I have some vague recollection that there might have been reasons not to. Do you remember @seallard ?

**Comment by islean** [comment]
Deployed to production:

```
Successfully installed cg-60.5.0
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/isak.ohlsson
done.
Log deploy... done.
remote: Enumerating objects: 701, done.
remote: Counting objects: 100% (701/701), done.
remote: Compressing objects: 100% (309/309), done.
remote: Total 701 (delta 484), reused 568 (delta 391), pack-reused 0
Receiving objects: 100% (701/701), 1.51 MiB | 0 bytes/s, done.
Resolving deltas: 100% (484/484), completed with 101 local objects.
From https://github.com/Clinical-Genomics/cg
   73c9725..dd40eb3  master     -> origin/master
   adf0304..677ae47  3092-allow-tomte-order -> origin/3092-allow-tomte-order
   cc4d759..5b663e1  block-FFPE-loqus-uploads -> origin/block-FFPE-loqus-uploads
   6ce2d4d..0771982  clean_past_run_dirs -> origin/clean_past_run_dirs
   13a6c76..a6d83dc  create-validation-sample -> origin/create-validation-sample
 + 76b6345...198fc1c different-external-apis -> origin/different-external-apis  (forced update)
 * [new branch]      fix-started-at-balsamic -> origin/fix-started-at-balsamic
 * [new branch]      remove-FFPE-from-loqusdb -> origin/remove-FFPE-from-loqusdb
 * [new tag]         v60.5.0    -> v60.5.0
Already on 'master'
Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
Updating 73c9725..dd40eb3
Fast-forward
 .bumpversion.cfg                       |   2 +-
 cg/__init__.py                         |   2 +-
 cg/cli/archive.py                      |  48 ++++++++++++++++++++++++++++++++++++++++++++++++
 cg/meta/archive/archive.py             |  31 +++++++++++++++++++++++++++----
 cg/meta/workflow/analysis.py           |  12 ++++++++----
 pyproject.toml                         |   2 +-
 tests/meta/archive/conftest.py         |  12 +++++++++++-
 tests/meta/archive/test_archive_api.py | 116 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 tests/meta/workflow/test_analysis.py   |   4 ++--
 9 files changed, 214 insertions(+), 15 deletions(-)
/home/isak.ohlsson
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
[isak.ohlsson@hasta:~] [P_base] $ 
```

**Pull request merged**